### PR TITLE
remote-write: Return 4xx when request contains a timeseries with duplicate labels.

### PR DESCRIPTION
**Pull request by dtrejod**
## Problem 

Came across an issue where grafana-agent writing to a Prometheus server and hitting a TSDB append error where the error was `label name "job" is not unique`. While a duplicate label is problematic within itself, I found it very strange that Prometheus itself was returning a 5xx error for this specific error. I believe the Prometheus spec calls for any 5xx error in a remote-write request to be retried by the client. This error case results in a scenario where the client continuously tries to send the bad time series until it eventually gives from queue_config/wal configuration.

`ErrInvalidSample` in TSDB for duplicate labels returned here: https://github.com/prometheus/prometheus/blob/main/tsdb/head_append.go#L434

## Proposal

For each RemoteWrite request (V1/V2) we already check if any time series has an invalid set of labels. In this PR I introduce an additional safety check for duplicate labels which itself will result in transparently swallowing the error (V1) or returning a 4xx error (V2).

### CODE REVIEW

File: storage/remote/write_handler_test.go [hunk 1]
```diff
@@ -338,6 +338,15 @@ func TestRemoteWriteHandler_V2Message(t *testing.T) {
...
 		},
+		{
+			desc: "Partial write; first series with duplicate labels",
+			input: append(
+				// Series with __name__="test_metric1",test_metric1="test_metric1",test_metric1="test_metric1" labels.
+				[]writev2.TimeSeries{{LabelsRefs: []uint32{1, 2, 2, 2, 2, 2}, Samples: []writev2.Sample{{Value: 1, Timestamp: 1}}}},
+				writeV2RequestFixture.Timeseries...),
+			expectedCode:     http.StatusBadRequest,
+			expectedRespBody: "invalid labels for series, labels {__name__=\"test_metric1\", test_metric1=\"test_metric1\", test_metric1=\"test_metric1\"}, duplicated label test_metric1\n",
+		},
```
Comment (dtrejod) [inline storage/remote/write_handler_test.go:349]: I only added a test for V2. In V1 the current API swallows any invalid label errors and returns a 204.

Happy to add an additional test for V1 if we want to test for asserting a 204 for duplicate label condition.

**Comment by bboreham** [comment]
I would expect to be able to send the same series at multiple different timestamps; is that allowed by this change?

**Comment by dtrejod** [comment]
> I would expect to be able to send the same series at multiple different timestamps; is that allowed by this change?

Maybe I can clarify what this PR fixes with the below example. If you send a RemoteWrite request with the following, it will result in a 5xx error instead of 4xx. 
```
{__name__="metric",job="foo",job="foo"}
```

To answer your question, no this PR will not prevent sending multiple samples for the same TimeSeries which does have a valid labelset (i.e no duplicate labels, no empty labels, ect).

**Review by cstyan — APPROVED** [review]
LGTM, as you pointed out this doesn't block having multiple timeseries with the same label set just a series which has an invalid label set via containing duplicate label names.
